packages/medusa/src/services/event-bus.ts

```diff
@@ -165,106 +194,60 @@ export default class EventBusService {
   async emit<T>(
     eventName: string,
     data: T,
-    options: EmitOptions = {}
-  ): Promise<StagedJob | void> {
-    if (this.transactionManager_) {
-      const stagedJobRepository = this.transactionManager_.getCustomRepository(
-        this.stagedJobRepository_
-      )
-
-      const stagedJobInstance = stagedJobRepository.create({
-        event_name: eventName,
-        data,
-      } as StagedJob)
-      return await stagedJobRepository.save(stagedJobInstance)
-    } else {
-      const opts: { removeOnComplete: boolean } & EmitOptions = {
-        removeOnComplete: true,
-      }
-      if (typeof options.attempts === "number") {
-        opts.attempts = options.attempts
-        if (typeof options.backoff !== "undefined") {
-          opts.backoff = options.backoff
-        }
-      }
-      if (typeof options.delay === "number") {
-        opts.delay = options.delay
-      }
-      this.queue_.add({ eventName, data }, opts)
+    options: EmitOptions & { cacheKey?: string } = {}
+  ): Promise<void> {
+    // construct options for the job
+    const opts: { removeOnComplete: boolean } & EmitOptions = {
+      removeOnComplete: true,
     }
-  }
-
-  startEnqueuer(): void {
-    this.shouldEnqueuerRun = true
-    this.enqueue_ = this.enqueuer_()
-  }
-
-  async stopEnqueuer(): Promise<void> {
-    this.shouldEnqueuerRun = false
-    await this.enqueue_
-  }
-
-  async enqueuer_(): Promise<void> {
-    while (this.shouldEnqueuerRun) {
-      const listConfig = {
-        relations: [],
-        skip: 0,
-        take: 1000,
+    if (typeof options.attempts === "number") {
+      opts.attempts = options.attempts
+      if (isDefined(options.backoff)) {
+        opts.backoff = options.backoff
       }
+    }
+    if (typeof options.delay === "number") {
+      opts.delay = options.delay
+    }
 
-      const stagedJobRepo = this.manager_.getCustomRepository(
-        this.stagedJobRepository_
-      )
-      const jobs = await stagedJobRepo.find(listConfig)
+    // If a unique id is passed to the method, we cache the event data
+    // instead of processing it immediately. As a consumer, you should
+    // process the events when appropriate e.g. when a transaction has
+    // committed. Use method `processCachedEvents`.
+    if (options?.cacheKey) {
+      const cachedEvents =
+        (await this.cacheService_.get<EventData[]>(options.cacheKey)) || []
 
-      await Promise.all(
-        jobs.map((job) => {
-          this.queue_
-            .add(
-              { eventName: job.event_name, data: job.data },
-              { removeOnComplete: true }
-            )
-            .then(async () => {
-              await stagedJobRepo.remove(job)
-            })
-        })
-      )
+      const job = { eventName, data, options: opts }
+      const updateEvents = [...cachedEvents, job]
```

Comment:
todo: updatedEvents eventually but it should be in the passed instead of present which make thing of an action